transport-native-epoll/src/main/java/io/netty/channel/epoll/TunTapPacketLogger.java

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright 2014 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.channel.epoll;
+
+import java.net.InetAddress;
+import java.util.Formatter;
+import java.util.Locale;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufUtil;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.logging.LogLevel;
+import io.netty.handler.logging.LoggingHandler;
+
+/**
+ * A {@link ChannelLogger} that logs packets sent or received via a {@link TunTapChannel}.
+ */
+public class TunTapPacketLogger extends LoggingHandler {
+
+    private boolean logPacketData = true;
+
+    public TunTapPacketLogger() {
+        super(LogLevel.INFO);
+    }
+
+    public TunTapPacketLogger(LogLevel level) {
+        super(level);
+    }
+
+    public TunTapPacketLogger(String name) {
+        super(name);
+    }
+
+    public TunTapPacketLogger(String name, LogLevel level) {
+        super(name, level);
+    }
+
+    public boolean logPacketData() {
+        return logPacketData;
+    }
+
+    public void setLogPacketData(boolean logPacketData) {
+        this.logPacketData = logPacketData;
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        if (msg instanceof TunTapPacket) {
+            logPacket((TunTapPacket) msg, ((TunTapChannel) ctx.channel()).isTapChannel(), "TUN/TAP PACKET RECEIVED");
+            ctx.fireChannelRead(msg);
+        } else {
+            super.channelRead(ctx, msg);
+        }
+    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+        if (msg instanceof TunTapPacket) {
+            logPacket((TunTapPacket) msg, ((TunTapChannel) ctx.channel()).isTapChannel(), "TUN/TAP PACKET SENT");
+            ctx.write(msg, promise);
+        } else {
+            super.write(ctx, msg, promise);
+        }
+    }
+
+    private void logPacket(TunTapPacket packet, boolean isEtherFrame, String eventName) {
+        if (logger.isEnabled(internalLevel)) {
+            logger.log(internalLevel, formatPacket(packet, isEtherFrame, eventName, logPacketData));
+        }
+    }
+
+    public static String formatPacket(TunTapPacket packet, boolean isEtherFrame, String title, boolean logPacketData) {
+        StringBuilder out = new StringBuilder();
```

Comment:
consider providing an estimate for the initial size ... seems likely to be bigger than the default of `16`.